tree/dataframe/src/RDFHelpers.cxx

```diff
@@ -57,3 +64,161 @@ void ROOT::RDF::RunGraphs(std::vector<RResultHandle> handles)
    for (auto &h : uniqueLoops)
       run(h);
 }
+
+
+ROOT::RDF::ProgressHelper::ProgressHelper(std::size_t increment,
+    std::size_t maxEvent,
+    unsigned int progressBarWidth,
+    unsigned int printInterval,
+    bool useShellColors) :
+fPrintInterval(printInterval),
+fMaxEvents{maxEvent},
+fIncrement{increment},
+fBarWidth{progressBarWidth},
+#ifdef _WIN32
+fIsTTY{_isatty(_fileno(stdout)) == 1},
+fUseShellColours{false && useShellColors}
+#else
+fIsTTY{isatty(fileno(stdout)) == 1},
+fUseShellColours{useShellColors && fIsTTY} // Control characters only with terminals.
+#endif
+{
+
+}
+
+/// Compute a running mean of events/s.
+double ROOT::RDF::ProgressHelper::EvtPerSec() const {
+  if (fEventsPerSecondStatisticsIndex < fEventsPerSecondStatistics.size())
+    return std::accumulate(fEventsPerSecondStatistics.begin(), fEventsPerSecondStatistics.begin() + fEventsPerSecondStatisticsIndex, 0.) / fEventsPerSecondStatisticsIndex;
+  else
+    return std::accumulate(fEventsPerSecondStatistics.begin(), fEventsPerSecondStatistics.end(), 0.) / fEventsPerSecondStatistics.size();
+}
+
+/// Record current event counts and time stamp, populate evts/s statistics array.
+std::pair<std::size_t, std::chrono::seconds>
+ROOT::RDF::ProgressHelper::RecordEvtCountAndTime() {
+  using namespace std::chrono;
+
+  const auto currentEventCount = fProcessedEvents.load();
+  const auto eventsPerTimeInterval = currentEventCount - fLastProcessedEvents;
+  fLastProcessedEvents = currentEventCount;
+
+  const auto oldPrintTime = fLastPrintTime;
+  fLastPrintTime = system_clock::now();
+
+  const duration<double> secondsCurrentInterval = fLastPrintTime - oldPrintTime;
+  fEventsPerSecondStatistics[fEventsPerSecondStatisticsIndex++ % fEventsPerSecondStatistics.size()] = eventsPerTimeInterval / secondsCurrentInterval.count();
+
+  return {currentEventCount, duration_cast<seconds>(fLastPrintTime - fBeginTime)};
```

Comment:
to reduce contention on "shared" resource (even though the other  access might only be reads).  
```suggestion
  const auto oldPrintTime = fLastPrintTime;
  const auto newPrintTime = system_clock::now();
  fLastPrintTime = system_clock::now();

  const duration<double> secondsCurrentInterval = newPrintTime - oldPrintTime;
  fEventsPerSecondStatistics[fEventsPerSecondStatisticsIndex++ % fEventsPerSecondStatistics.size()] = eventsPerTimeInterval / secondsCurrentInterval.count();

  return {currentEventCount, duration_cast<seconds>(newPrintTime - fBeginTime)};
```